mlos_core/mlos_core/optimizers/optimizer.py

```diff
@@ -73,201 +75,182 @@ def space_adapter(self) -> Optional[BaseSpaceAdapter]:
         # Do some input validation.
-        assert metadata is None or isinstance(metadata, pd.DataFrame)
-        assert set(scores.columns) == set(
+        assert observation.metadata is None or isinstance(observation.metadata, pd.Series)
+        assert set(observation.score.index) == set(
             self._optimization_targets
         ), "Mismatched optimization targets."
         assert self._has_context is None or self._has_context ^ (
-            context is None
+            observation.context is None
         ), "Context must always be added or never be added."
-        assert len(configs) == len(scores), "Mismatched number of configs and scores."
-        if context is not None:
-            assert len(configs) == len(context), "Mismatched number of configs and context."
-        assert configs.shape[1] == len(
+        assert len(observation.config) == len(
             self.parameter_space.values()
         ), "Mismatched configuration shape."
-        self._observations.append((configs, scores, context))
-        self._has_context = context is not None
 
+        self._has_context = observation.context is not None
+        self._observations.append(observation)
+
+        register_observation = deepcopy(observation)  # Needed to support named tuples
         if self._space_adapter:
-            configs = self._space_adapter.inverse_transform(configs)
-            assert configs.shape[1] == len(
+            register_observation.config = self._space_adapter.inverse_transform(
+                register_observation.config
+            )
+            assert len(register_observation.config) == len(
                 self.optimizer_parameter_space.values()
             ), "Mismatched configuration shape after inverse transform."
-        return self._register(configs=configs, scores=scores, context=context)
+
+        return self._register(observation=register_observation)
 
     @abstractmethod
     def _register(
         self,
         *,
-        configs: pd.DataFrame,
-        scores: pd.DataFrame,
-        context: Optional[pd.DataFrame] = None,
-        metadata: Optional[pd.DataFrame] = None,
+        observation: Observation,
     ) -> None:
         """
         Registers the given configs and scores.
 
         Parameters
         ----------
-        configs : pd.DataFrame
-            Dataframe of configs / parameters. The columns are parameter names and
-            the rows are the configs.
-        scores : pd.DataFrame
-            Scores from running the configs. The index is the same as the index of the configs.
-
-        context : pd.DataFrame
-            Not Yet Implemented.
+        observation: Observation
+            The observation to register.
         """
         pass  # pylint: disable=unnecessary-pass # pragma: no cover
 
     def suggest(
         self,
         *,
-        context: Optional[pd.DataFrame] = None,
+        context: Optional[pd.Series] = None,
         defaults: bool = False,
```

Comment:
Can we make context a `data_type` too?

Reply:
I'm not sure that this makes sense, at least for now. Context is exclusively a series.

Reply:
It's a little weird since context becomes a dataframe in Observations. Also if we did this, we should also do it for metadata, and probably also config, both of which also have this problem that in Observations they go from series to dataframes.

Reply:
I have a minimal implementation in the PR now, but really not sure what to do with observations, other than take a list of series, which feels pretty awkward

Reply:
@bpkroth

Reply:
K, fine, but then let's leave the '*' param so that it's always called as

`opt.suggest(context=my_context_series)`